### PR TITLE
Week2

### PR DESCRIPTION
**Pull request by JuanOnOne**
Week 2 submission. 

## Release Notes

- **New Features**
  - Introduced multiple Databricks notebooks for data preparation, machine learning experiments, model logging, and feature engineering.
  - Added a new `DataProcessor` class for efficient data handling and preprocessing.
  - Implemented a `CustomModelWrapper` for custom model predictions and logging.
  - Enhanced model training and logging capabilities with MLflow integration.
  - Added a new function for adjusting predictions based on a scale factor.

- **Bug Fixes**
  - Removed unnecessary whitespace in the README for improved readability.

- **Chores**
  - Updated dependency management with specific version pinning in `pyproject.toml`.
  - Adjusted the CI workflow for consistent dependency installation.
  - Simplified project configuration by removing certain parameters.

- **Documentation**
  - Enhanced project configuration management with a new `ProjectConfig` class.

### CODE REVIEW

File: notebooks/05.log_and_register_fe_model.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+# Define table names and function name
+feature_table_name = f"{catalog_name}.{schema_name}.bank_marketing_features"
+
+# COMMAND ----------
+# Load training and test sets
+# Load training and test sets
+train_set = spark.table(f"{catalog_name}.{schema_name}.train_set")
+features_renamed = [f.replace(".", "_") if "." in f else f for f in train_set.columns]
+train_set = train_set.toDF(*features_renamed)
```
Comment (mvechtomova) [inline notebooks/05.log_and_register_fe_model.py]: this could be problematic for serving, I am afraid. the original train set must have the renamed features already - I have not tried this way, but if serving fails (can not find column names, and I am almost certain that would be the issue) this would be the reason.

File: notebooks/05.log_and_register_fe_model.py [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+training_df = training_set.load_df().toPandas()
+training_df.rename(columns={"default_": "default"}, inplace=True)
+
+# Calculate house_age for training and test set
+# test_set = spark.createDataFrame(test_set)
+test_set = test_set.selectExpr("*", f"{function_name}(default, housing, loan) as risk")
+test_set = test_set.toPandas()
+
+# Split features and target
+num_features = [f.replace(".", "_") for f in num_features]
```
Comment (mvechtomova) [inline notebooks/05.log_and_register_fe_model.py]: this will definitely be a problem

File: src/mlops_with_databricks/preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        self.X = None
+        self.y = None
+        self.preprocessor = None
+        self.target_encoder = None
+
+    def load_data(self, filepath):
+        df = self.spark.read.option("delimiter", ";").option("header", "true").csv(filepath).toPandas()
+        return df
+
+    def prepare_data(self):
```
Comment (mvechtomova) [inline src/mlops_with_databricks/preprocessor.py:28]: all the renaming must be done here

**Review by mvechtomova — COMMENTED** [review]
nice job! I see an issue with feature renaming after you created the train set, this will be problematic for serving.

File: notebooks/05.log_and_register_fe_model.py (hunk 1 above)
Comment (JuanOnOne) [reply notebooks/05.log_and_register_fe_model.py]: Thanks @mvechtomova  for taking a look. If you run this with "." in the column names it was failing, providing with very cryptic errors. 
Line 151 in my code was failing if I didn't replace "." in the columns, which is how the data came in from the source: 
`training_df = training_set.load_df().toPandas()`

**Comment by JuanOnOne** [comment]
@mvechtomova I've made the changes to preprocessor function and the fe model. Let me know!

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -11,15 +11,24 @@ jobs:
...
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.11'
 
-      - name: Install the dependencies
-        run: uv sync 
+      - name: Create and activate virtual environment
+        run: |
+          python -m venv .venv
+          echo "VIRTUAL_ENV=${{ github.workspace }}/.venv" >> $GITHUB_ENV
```
Comment (mvechtomova) [inline .github/workflows/ci.yml:22]: I do not think it makes much sense to create a virtul env. You install python from scratch, no point in doing this.

File: notebooks/05.log_and_register_fe_model.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+# Set and start MLflow experiment
+mlflow.set_experiment(experiment_name="/Shared/bank-marketing-fe")
+git_sha = "ffa63b430205ff7"
+
+with mlflow.start_run(tags={"branch": "week2", "git_sha": f"{git_sha}"}) as run:
+    run_id = run.info.run_id
+
+    # Pipeline fit/predict
+    pipeline.fit(X_train, le.fit_transform(y_train))
```
Comment (mvechtomova) [inline notebooks/05.log_and_register_fe_model.py:201]: you create the label encoder, but do not save the mapping anywhere. So you run into problems in the future, should not be that part of preprocessing?

**Review by mvechtomova — COMMENTED** [review]
This label encoder is problematic. You do not save the results of that mapping anywehere. I think ideally you should create an extra column in train and test set with the result of that encoding, and use it later to output the prediction.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (JuanOnOne) [reply .github/workflows/ci.yml:22]: @mvechtomova what should I do then if I'm using venv?
